### ISSUE TITLE
TexLive: xindy is not available

### ISSUE THREAD

**Issue by jokogr** [NixOS/nixpkgs#16207]
## Issue description

The `xindy` command is not available.

@vcunat any idea what's missing?

### Steps to reproduce

Install `xindy` or `collection-binextra`, e.g.:

```
    (texlive.combine {
     inherit (texlive) scheme-basic collection-bibtexextra collection-binextra
     collection-fontsrecommended collection-genericextra
     collection-genericrecommended collection-langgreek collection-latex
     collection-latexextra collection-latexrecommended collection-metapost
     collection-publishers collection-science collection-xetex xindy;
    })
```

```
 % xindy
zsh: command not found: xindy
```

## Technical details

* System: NixOS 16.09.git.c3d5a0a (Flounder)
* Nix version: nix-env (Nix) 1.11.2
* Nixpkgs version: 16.09.git.c3d5a0a

**Comment by vcunat** [comment]
xindy wasn't packaged yet for nixpkgs (the binary part). I originally left it out because it needed nontrivial extra work, as noted on http://tug.org/texlive/build.html#xindy

**Comment by vcunat** [comment]
@jokogr: I added a binary in 97325d7 (i.e. not in official repo yet) &ndash; will you test that it works well? I'm not familiar with `xindy` so I just made `xindy --help` work.

**Comment by jokogr** [comment]
@vcunat Many thanks for your really fast work!

`xindy --help` works here too, but when I am running it for real work I am getting this message:

```
% latexmk thesis.tex
Latexmk: This is Latexmk, John Collins, 5 February 2015, version: 4.43a.
Latexmk: applying rule 'cusdep glo gls thesis'...
Rule 'cusdep glo gls thesis': File changes, etc:
   Non-existent destination files:
      'thesis.gls'
------------
Run number 1 of rule 'cusdep glo gls thesis'
------------
For rule 'cusdep glo gls thesis', running '&do_cusdep( run_makeglossaries )' ...
makeglossaries version 2.18 (2016-01-24)
added glossary type 'main' (glg,gls,glo)
added glossary type 'acronym' (alg,acr,acn)
xindy  -L greek -C utf8 -I xindy -M "thesis" -t "thesis.glg" -o "thesis.gls" "thesis.glo"
xindy: Cannot locate /nix/store/ajjk5c0hy5j8snmba3bvrpb4jp56i61k-texlive-xindy.bin-2015/lib/xindy/xindy.mem at /nix/store/ajjk5c0hy5j8snmba3bvrpb4jp56i61k-texlive-xindy.bin-2015/bin/xindy line 519.

***Call to xindy failed***

Check 'thesis.glg' for details
Rule 'cusdep glo gls thesis', function 'run_makeglossaries'
   failed with return code = 6400
Latexmk: Errors, so I did not complete making targets
Collected error summary (may duplicate other messages):
  cusdep glo gls thesis: Command for 'cusdep glo gls thesis' gave return code 6400
Latexmk: Use the -f option to force complete processing,
 unless error was exceeding maximum runs of latex/pdflatex.
```

P.S. (out of topic): Pardon for my ignorance, is there a fast way to get this commit of yours directly in my repo? I had to copy-paste it.

**Comment by vcunat** [comment]
> P.S. [...]

Yes, github provides an easy way, as many URLs can be used with ".diff" or ".patch" suffix:
```
$ curl https://github.com/NixOS/nixpkgs/commit/97325d7814.patch | git am
```

**Comment by vcunat** [comment]
Another error should be fixed up in d0eb0ef (replacing the previous patch). I don't get why it's trying to locate files on different paths than it installed to...

**Comment by jokogr** [comment]
Thanks for the github tip!

Here's what I'm getting now:

```
% latexmk thesis.tex
Latexmk: This is Latexmk, John Collins, 5 February 2015, version: 4.43a.
Latexmk: applying rule 'cusdep glo gls thesis'...
Rule 'cusdep glo gls thesis': File changes, etc:
   Non-existent destination files:
      'thesis.gls'
------------
Run number 1 of rule 'cusdep glo gls thesis'
------------
For rule 'cusdep glo gls thesis', running '&do_cusdep( run_makeglossaries )' ...
makeglossaries version 2.18 (2016-01-24)
added glossary type 'main' (glg,gls,glo)
added glossary type 'acronym' (alg,acr,acn)
xindy  -L greek -C utf8 -I xindy -M "thesis" -t "thesis.alg" -o "thesis.acr" "thesis.acn"
xindy: Could not execute xindy kernel: No such file or directory

***Call to xindy failed***

Check 'thesis.alg' for details
Rule 'cusdep glo gls thesis', function 'run_makeglossaries'
   failed with return code = 6400
Latexmk: Errors, so I did not complete making targets
Collected error summary (may duplicate other messages):
  cusdep glo gls thesis: Command for 'cusdep glo gls thesis' gave return code 6400
Latexmk: Use the -f option to force complete processing,
 unless error was exceeding maximum runs of latex/pdflatex.
```

**Comment by vcunat** [comment]
Hmm, I'd think that `xindy.mem` would be the core/kernel, but it seems I'm mistaken. I know little about lisp implementation details.

**Comment by jokogr** [comment]
@vcunat no, I think you solved it! If I include clisp to my systemPackages, it works. I bet xindy could not find clisp when running.

P.S.: You could use `xindy -V` instead of `--help` to check if it works.

**Comment by vcunat** [comment]
It would be better if we managed to hard-code the clisp path into xindy.

**Comment by jokogr** [comment]
My guess is that a shebang needs to be patched for clisp.

Could you help me getting to the build phase? I've tried with

```
nix-shell -E 'with import <nixpkgs> {}; texlive-xindy.bin-2015'
```

or

```
nix-shell -E 'with import <nixpkgs> {}; texlive.combine { inherit(texlive) scheme-basic xindy; }'
```

but no luck :(

**Comment by jokogr** [comment]
ok, my nix-fu is limited, so I had to extract the sources manually and experiment with them. I have found out that xindy (as a Perl script) invokes at some point clisp, so the relevant line in the xindy script needs to be changed.

I made it work in my NixOS installation by adding the following in the postInstall step (thanks to obadz and clever from Freenode #nixos):

```
substituteInPlace "$out"/bin/xindy --replace "'clisp'" "'${clisp}/bin/clisp'"
```

Just be careful to maintain the single quotes ("'clisp'" vs "clisp"), there are multiple clisp in xindy, but only one 'clisp'.
